### PR TITLE
APIGOV-22558 - Support for provisioning OAuth client to IDP

### PR DESCRIPTION
**Pull request by vivekschauhan**
- Changes to bind object array based properties to environment variable based config
- Changes to parse IDP config from environment based config
- Added OAuth provider registry to maintain registered/configured providers
- Added OAuth provider to manage the registration/un-registration of the OAuth clients based on OAuth dynamic client registration protocol
- Provider are initialized fetching the OAuth authorization server metadata
- Currently added support for generic(which includes Keycloak) and Okta
- Added auth client to acquire token using client/secret or public/private keypair based authenticator
- Refactored apic/auth platform token getter to reuse the auth client with public/private keypair authenticator
- Setup CRD for IDP provider with request provision schema for the data needed to provision external OAuth client with IDP
- Changes to credential request interface to identify if the request is for IDP and provide data for the provisioned OAuth client
- Changes to register OAuth client before calling CredentialProvision handler. The data for registered OAuth client is passed to the handler to allow agent implementation to associate the information to dataplane application
- Change to deregister OAuth client after CredentialDeprovision handler

### CODE REVIEW

**Review by sbolosan — APPROVED** [review]

File: pkg/agent/agent.go [hunk 1]
```diff
@@ -198,6 +201,25 @@ func InitializeWithAgentFeatures(centralCfg config.CentralConfig, agentFeaturesC
...
+		for _, idp := range idPCfg.GetIDPList() {
+			registerCredentialProvider(idp, tlsCfg, proxy, timeout)
+		}
+	}
+}
+
+func registerCredentialProvider(idp config.IDPConfig, tlsCfg config.TLSConfig, proxyURL string, clientTimeout time.Duration) {
+	err := GetAuthProviderRegistry().RegisterProvider(idp, tlsCfg, proxyURL, clientTimeout)
+	if err != nil {
+		log.Errorf("unable to register external IdP provider, any credential request to the IdP will not be processed. %s", err.Error())
```
Comment (tsjohns9) [inline pkg/agent/agent.go:219]: Can you please use the logger in this file

File: pkg/agent/handler/credential.go [hunk 2]
```diff
@@ -194,67 +234,164 @@ func (h *credentials) getCRD(ctx context.Context, cred *mv1.Credential) (*mv1.Cr
...
+	return provCred, nil
+}
+
+// newIDPCredData - reads the idp client metadata from credential request
+func newIDPCredData(p oauth.Provider, credData, provData map[string]interface{}) *idpCredData {
+	cd := &idpCredData{}
+
+	if provData != nil {
+		if data, ok := provData[prov.OauthClientID]; ok && data != nil {
+			cd.clientID = data.(string)
```
Comment (tsjohns9) [inline pkg/agent/handler/credential.go:343]: Maybe check if the type conversion was ok

Comment (tsjohns9) [reply pkg/agent/handler/credential.go:343]: It's up to you. There are a few places below here too that don't check for ok. But maybe it is fine.

File: pkg/agent/handler/credential_test.go [hunk 3]
```diff
@@ -138,7 +138,7 @@ func TestCredentialHandler(t *testing.T) {
 				subError:       tc.subError,
 			}
 
-			handler := NewCredentialHandler(p, c)
+			handler := NewCredentialHandler(p, c, nil)
```
Comment (tsjohns9) [inline pkg/agent/handler/credential_test.go:141]: Can you please add a test with the oauth provider registry?

File: pkg/authz/oauth/authclient.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+	cachedTokenExpiry *time.Timer
+	getTokenMutex     *sync.Mutex
+	options           *authClientOptions
+}
+
+type authenticator interface {
+	prepareRequest() (url.Values, error)
+}
+
+type tokenResponse struct {
```
Comment (tsjohns9) [inline pkg/authz/oauth/authclient.go:44]: The fields are exported, but the type is not. Is that intended?

File: pkg/authz/oauth/authclient.go [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+	}
+
+	resp, err := c.postAuthForm(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.Code != 200 {
+		body := resp.Body
+		log.Debugf("bad response from %s: %d %s: %s, request time : %s", c.options.serverName, resp.Code, http.StatusText(resp.Code), body, startTime.String())
```
Comment (tsjohns9) [inline pkg/authz/oauth/authclient.go:155]: Can you update to use the FieldLogger please

File: pkg/agent/handler/credential_test.go (hunk 3 above)
Comment (vivekschauhan) [reply pkg/agent/handler/credential_test.go:141]: Yes, still need to add few more tests

File: pkg/agent/handler/credential.go (hunk 2 above)
Comment (vivekschauhan) [reply pkg/agent/handler/credential.go:343]: Sure, will do that change

File: pkg/agent/agent.go (hunk 1 above)
Comment (vivekschauhan) [reply pkg/agent/agent.go:219]: Will do the change

File: pkg/authz/oauth/authclient.go (hunk 5 above)
Comment (vivekschauhan) [reply pkg/authz/oauth/authclient.go:155]: Sure

File: pkg/authz/oauth/provider.go [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+		metadataURL:     idp.GetMetadataURL(),
+		cfg:             idp,
+		extraProperties: idp.GetExtraProperties(),
+		apiClient:       apiClient,
+		idpType:         idpType,
+	}
+	metadata, err := p.fetchMetadata()
+	if err != nil {
+		p.logger.
+			WithField("name", p.cfg.GetIDPName()).
```
Comment (tsjohns9) [inline pkg/authz/oauth/provider.go:74]: Can you please add `WithError(err)`

File: pkg/util/testdata/private_key.pem [hunk 7]
```diff
@@ -0,0 +1,28 @@
+-----BEGIN PRIVATE KEY-----
```
Comment (tsjohns9) [inline pkg/util/testdata/private_key.pem:1]: Same here

**Review by tsjohns9 — APPROVED** [review]

File: pkg/util/testdata/password [hunk 8]
```diff
@@ -0,0 +1 @@
+hammertime
```
[the remark this answers is not in the document]
Comment (vivekschauhan) [reply pkg/util/testdata/password:1]: This is test data (already being used by apic/auth package)

File: pkg/util/testdata/private_key.pem (hunk 7 above)
Comment (vivekschauhan) [reply pkg/util/testdata/private_key.pem:1]: This is test data (already being used by apic/auth package)

**Review by sbolosan — DISMISSED** [review]

**Review by sbolosan — DISMISSED** [review]

**Review by sbolosan — DISMISSED** [review]

**Review by sbolosan — APPROVED** [review]

**Review by jcollins-axway — APPROVED** [review]

**Review by sbolosan — APPROVED** [review]

**Pull request merged**